python/guide/modules/autograd.md

```diff
@@ -1,178 +1,260 @@
+        <source src="./imgs/autograd_grad_req.mp4" type="video/mp4">
+    </video>
+</p>
+
+<a name="nonparamgrad"></a>
+## Advanced: Calculating non-parameter gradients
+
+Although it's most common to deal with network parameters (with `Parameter` being an MXNet Gluon abstraction), there are cases when you need to calculate the gradient with respect to thing that are not parameters (i.e. standard `ndarray`s). One example would be finding the gradient of the loss with respect to the input data to generate adversarial examples.
+
+With `autograd` it's simple, but there's one key difference compared to parameters: parameters are assumed to require gradients by default, non-parameters are not. We need to explicitly state that we require the gradient, and we do that by calling `.attach_grad()` on the `ndarray`. We can then access the gradient using `.grad` after the `backward` pass.
+
+As a simple example, let's take the case where $y=2x^2$ and use `autograd` to calculate gradient of $y$ with respect to $x$ at three different values of $x$. We could obviously work out the gradient by hand in this case as $dy/dx=4x$, but let's use this knowledge to check `autograd`. Given $x$ is an `ndarray` and not a `Parameter`, we need to call `x.attach_grad()`.
 
 ```{.python .input  n=9}
-a = nd.random.normal(shape=1)
-a.attach_grad()
+x = array([1, 2, 3])
+x.attach_grad()
 with autograd.record():
-    d = f(a)
-d.backward()
+    y = 2 * x ** 2
+y.backward()
+print(x.grad)
 ```
 
-Let's analyze the `f` function defined above. As you can see, it is piecewise
-linear in its input `a`. In other words, for any `a` there exists some constant
-such that for a given range `f(a) = g * a`. Consequently `d / a` allows us to
-verify that the gradient is correct:
+<a name="controlflow"></a>
+## Advanced: Using Python control flow
+
+As mentioned before, one of the main advantages of `autograd` is the ability to automatically calculate gradients of dynamic graphs (i.e. graphs where the operators could be different on every forward pass). One example of this would be applying a tree structured recurrent network to parse a sentence using its parse tree. And we can use Python control flow operators to create a dynamic flow that depends on the data.
+
+Warning: you can't `hybridize` Python control flow, to take advantage of symbolic graph optimizations, but you can use [Control Flow Operators](https://mxnet.incubator.apache.org/versions/master/tutorials/control_flow/ControlFlowTutorial.html) instead.
+
+We'll write a function as a toy example of a dynamic network. We'll add an `if` condition and a loop with a variable number of iterations, both of which will depend on the input data. Although these can now be used in static graphs (with conditional operators) it's still much more natural to use native control flow.
+
 
 ```{.python .input  n=10}
-print(a.grad == (d / a))
+import math
+
+def f(x):
+    # going to change y but still want reference to x
```

Comment:
but still want to store x (there's no reference, x is a primitive type)

Reply:
changed to 'going to change y but still want to use x'. but will get you to explain this one!